pkg/syncer/util.go

```diff
@@ -422,9 +425,65 @@ func getVcHostAndVolumeManagerForVolumeID(ctx context.Context,
 			return "", nil, logger.LogNewErrorf(log,
 				"could not get volume manager for the vCenter: %q", vCenter)
 		}
+
+		log.Debugf("Identified VC %s for multi VC setup for volume %s", vCenter, volumeID)
 		return vCenter, volumeManager, nil
 	}
 
 	return "", nil, logger.LogNewErrorf(log,
 		"failed to get VC host and volume manager. VolumeInfoService is not initialized.")
 }
+
+// getVcHostAndVolumeManagerFromTopologySegments returns VC host and the corresponding
+// volume manager that can access the given volume on VC.
+// Based on the nodeAffinity rules defined on the PV, this function finds out
+// the right VC. If overlapping topology segments are found in nodeAffinity,
+// error is returned.
+func getVcHostAndVolumeManagerFromTopologySegments(ctx context.Context, pv *v1.PersistentVolume, metadataSyncer *metadataSyncInformer) (string, volumes.Manager, error) {
+	log := logger.GetLogger(ctx)
+	var vcHost string
+
+	log.Debugf("Getting VC from topology segments for volume %s", pv.Name)
+
+	if pv.Spec.NodeAffinity != nil {
+		if pv.Spec.NodeAffinity.Required != nil {
+			if pv.Spec.NodeAffinity.Required.NodeSelectorTerms != nil {
+				for _, nodeSelector := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+					if nodeSelector.MatchExpressions == nil {
+						continue
+					}
+					// Get topology segments on PV
+					topologySegments := make(map[string]string)
+					for _, topology := range nodeSelector.MatchExpressions {
+						for _, val := range topology.Values {
+							topologySegments[topology.Key] = val
+						}
+					}
+
+					vc, err := getVCForTopologySegments(ctx, topologySegments)
```

Comment:
Can you prepare a complete list of all segments from all node affinity rules and then invoke `getVCForTopologySegments`